packages/compiler/core/util.ts

```diff
@@ -322,3 +322,102 @@ class CaseInsensitiveMap<T> extends Map<string, T> {
     return super.delete(key.toUpperCase());
   }
 }
+
+interface OrderedMapKey {
+  key: string;
+}
+
+export class OrderedMap<V> implements Map<string, V> {
```

Comment:
Could technically also have this take the `K` template argument. The key doesn't need to be a string for this to work. Don't think we'll have a use for it but it should work.

Reply:
I think using the same type signature as map will be a win just from a readability standpoint.